odl/solvers/functional/default_functionals.py

```diff
@@ -570,6 +573,467 @@ def __init__(self, space):
+    The functional is given by
+
+    .. math::
+
+        \\sum_{i} \left( g_i - x_i + x_i \log \left( \\frac{x_i}{ pos(g_i) }
+        \\right) \\right) + I_{x \\geq 0}(x)
+
+    where :math:`g` is the prior, and :math:`I_{x \\geq 0}(x)` is the indicator
+    function on nonnegative elements.
+
+    Mention something about the use of this, and compare to
+    KullbackLeibler.
+
+    See Also
+    --------
+    KullbackLeiblerCrossEntropyConvexConj : the convex conjugate functional
+    KullbackLeibler : related functional
+    """
+
+    # TODO: update doc above
+
+    def __init__(self, space, prior=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        prior : ``space`` element, optional
+            Data term, positive. If None it is take as the one-element.
+        """
+        super().__init__(space=space, linear=False, grad_lipschitz=np.nan)
+
+        if prior is not None and prior not in self.domain:
+            raise ValueError('`prior` not in `domain`'
+                             ''.format(prior, self.domain))
+
+        self.__prior = prior
+
+    @property
+    def prior(self):
+        """The prior in the Kullback-Leibler functional."""
+        return self.__prior
+
+    # TODO: update when integration operator is in place: issue #440
+    def _call(self, x):
+        """Return the KL-diveregnce in the point ``x``.
+
+        If any point in x is zero or smaller, the value is infinite.
+        """
+        if self.prior is None:
+            tmp = ((1 - x + x * np.log(x)).inner(self.domain.one()))
+        else:
+            tmp = ((self.prior - x + x * np.log(x / self.prior))
+                   .inner(self.domain.one()))
+        if np.isnan(tmp):
+            # In this case, some element was less than or equal to zero
+            return np.inf
+        else:
+            return tmp
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        The gradient is not defined in points where one or more components
+        are less than or equal to 0.
+        """
+        functional = self
+
+        class KLCrossEntropyGradient(Operator):
+
+            """The gradient operator of this functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Apply the gradient operator to the given point.
+
+                The gradient is not defined in for points with components less
+                than or equal to zero.
+                """
+                if functional.prior is None:
+                    tmp = np.log(x)
+                else:
+                    tmp = np.log(x / functional.prior)
+
+                if any(np.isnan(tmp)) or -np.inf in tmp:
```

Comment:
I use the `nan`-test to catch if any `x` is `< 0` and the `-inf`-test to catch if any `x` is = 0 (`np.log` is defined differently in these cases).